### PR TITLE
Migrate license mismatch allowlist to Homebrew/core

### PR DESCRIPTION
**Pull request by Rylan12**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally? (I've run the audit tests only)
- [ ] Have you successfully run `brew man` locally and committed any changes? (no changes)

-----

This PR migrates the `PERMITTED_FORMULA_LICENSE_MISMATCHES` audit allowlist to Homebrew/core.

Additionally, I added some more tests for the lists migrated in https://github.com/Homebrew/brew/pull/9039

Prompted by https://github.com/Homebrew/homebrew-core/pull/59828

Corresponding Homebrew/core PR (should be merged at the same time or before): https://github.com/Homebrew/homebrew-core/pull/65061

### CODE REVIEW

File: Library/Homebrew/test/dev-cmd/audit_spec.rb
Comment (reitermarkus) [inline Library/Homebrew/test/dev-cmd/audit_spec.rb:426]: *differs

Comment (MikeMcQuaid) [inline Library/Homebrew/test/dev-cmd/audit_spec.rb:426]:
```suggestion
      it "allows a formula-specified license that differs from its GitHub "\
```

**Review by MikeMcQuaid — APPROVED** [review]
👍🏻 to merge after typo resolved.

**Comment by Rylan12** [comment]
Thanks, both!

I'm going to spend some time reworking the logic here now that https://github.com/Homebrew/brew/pull/9187 has been merged. I'm planning on extracting the logic that handles the lists to a new location so it can be used by other modules as well (e.g. PyPI for https://github.com/Homebrew/brew/pull/9185).

Thoughts on:
1. Renaming `audit_exceptions` to e.g. `tap_lists` for `formula_lists`
1. The location for this code. I'm leaning toward `Library/Homebrew/tap_lists.rb` (or `formula_lists`, depending on number 1) unless another location makes more sense.

**Comment by MikeMcQuaid** [comment]
> * Renaming `audit_exceptions` to e.g. `tap_lists` for `formula_lists`

I prefer keeping `audit_exceptions` for the stuff related to audits and having different names/folders for things used for different tools. A more generic solution may mean less code but it makes the contributor/user workflow for figuring out what these files are for a little harder.

**Comment by Rylan12** [comment]
Okay, that's fair. How about going with directories that end in `*_lists/`. So rename the audit lists to `audit_exception_lists/` and then we can add others like `pypi_lists/`?

I'm trying to reduce the number of hardcoded files so adding lists in the future is pretty easy.

**Comment by Rylan12** [comment]
Okay, I've reworked the handling of lists a little bit.

Now, all directories that end in `*_lists` will be parsed and can be accessed via `tap.formula_lists`.

Tests are failing because this no longer picks up the lists from Homebrew/core. When this PR is ready, I'll merge https://github.com/Homebrew/homebrew-core/pull/65061 which will then need to be merged into linuxbrew-core.

**Comment by MikeMcQuaid** [comment]
> Now, all directories that end in `*_lists` will be parsed and can be accessed via `tap.formula_lists`.

That's a bit too magic for my liking. `audit_exceptions/` feels like a more appropriate name, too. 

> Okay, that's fair. How about going with directories that end in `*_lists/`. So rename the audit lists to `audit_exception_lists/` and then we can add others like `pypi_lists/`?
> I'm trying to reduce the number of hardcoded files so adding lists in the future is pretty easy.

Do we anticipate having more than one list for PyPI? If not: [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it).

Sorry to keep jerking you around here. I'm happy for you to not make any code changes until we're agree on the implementation (if that works for you)? I don't mind you making changes back and forth but I don't want to waste your time (if you feel like I am).

**Comment by Rylan12** [comment]
> That's a bit too magic for my liking. `audit_exceptions/` feels like a more appropriate name, too.

Okay, I think that's fair. Right now there are [some constants](https://github.com/Homebrew/brew/blob/181baaafb0107205d29d64bb477c13e34a9c214f/Library/Homebrew/tap.rb#L21-L29) in `tap.rb` so I think this would be a good place to hardcode which files we're going to allow. I was trying to make things work nicely, especially for third-party taps, but I'm not really sure it matters because the brew code will need to be updated anyway if new lists are going to be added.

> Do we anticipate having more than one list for PyPI? If not: [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it).

Probably not so fair point.

> Sorry to keep jerking you around here. I'm happy for you to not make any code changes until we're agree on the implementation (if that works for you)? I don't mind you making changes back and forth but I don't want to waste your time (if you feel like I am).

No worries! Not a waste of time. Timezones are a little tricky here because most of my free time is after the European maintainers go to bed, it seems, so I'm just trying to have workable solutions in place for the morning 😄

---

So, moving forward, are we thinking:
- We'll hardcode in `tap.rb` which files are the list files.
- We'll continue to use `audit_exceptions/` for the audit exception lists

So then, my question:
- Where will files that don't belong in the audit exceptions category live? The tap's root directory? A `formula_lists/` directory? Right now we only have one that I'm in the process of moving for pypi resources, but yesterday I saw another list (I believe relating to `livecheck` that won't fit under the `audit_exceptions` title. Maybe I need to do some more investigation into what lists we have that will be migrated so I can get a better sense of how many non-audit-exception-lists we're talking about.

**Comment by MikeMcQuaid** [comment]
>  but I'm not really sure it matters because the brew code will need to be updated anyway if new lists are going to be added.

Agreed 👍🏻 



> * I believe relating to `livecheck` that won't fit under the `audit_exceptions` title. Maybe I need to do some more investigation into what lists we have that will be migrated so I can get a better sense of how many non-audit-exception-lists we're talking about.

Yeh, I think this is probably required before we can figure out whether it's best to have one-or-more new directories or one-or-more new files in the repo root.

> * We'll hardcode in `tap.rb` which files are the list files.
> * We'll continue to use `audit_exceptions/` for the audit exception lists

Yup 👍🏻

**Comment by MikeMcQuaid** [comment]
> > Do we anticipate having more than one list for PyPI? If not: [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it).
> 
> Probably not so fair point.

Sorry, missed this bit. I'm more pointing to the principle here; I think it's easy enough to change these when we need them more rather than pre-planning things just in case we need them in future.

**Comment by Rylan12** [comment]
I should have some time in the next couple of hours to look into this a little more. I'll try to find all of the non-audit lists that will need to be migrated eventually and then, maybe, we can make a better decision (I'll need to find these lists in the future, anyway, so may as well do it now).

**Comment by MikeMcQuaid** [comment]
Thanks again @Rylan12, you're doing great work ❤️

**Comment by Rylan12** [comment]
Okay, I've looked through and found lists that should be migrated in the following categories:

- audit lists
    - We will keep using `audit_exceptions/`
- rubocop lists
    - We can either add these to the existing `audit_exceptions/` directory or create a new `rubocop_exceptions/` or `style_exceptions/` directory
- other lists
    - The `Livecheck` module's [`GITHUB_SPECIAL_CASES`](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/livecheck/livecheck.rb)
    - The `PyPI` module's [`AUTOMATIC_RESOURCE_UPDATE_BLOCKLIST`](https://github.com/Homebrew/brew/blob/ca333a5da0f1305671c9d701d69b94aa29e2f16e/Library/Homebrew/utils/pypi.rb#L13-L22)
    - We can either add these as individual files in the tap's root directory or can nest them under an e.g. `formula_lists/` directory. I'm leaning toward the second option because I think it is cleaner and adding more lists in the future won't increase clutter.

**Comment by MikeMcQuaid** [comment]
> * We will keep using `audit_exceptions/`

👍🏻 



> * We can either add these to the existing `audit_exceptions/` directory or create a new `rubocop_exceptions/` or `style_exceptions/` directory. I'm leaning toward `style_exceptions/` because it matches with the command (`brew style` not `brew rubocop`) so it's clearer to anyone who doesn't know about rubocop or the inner workings of `brew style`.

👍🏻 to either `audit_exceptions` or `style_exceptions`.



> * We can either add these as individual files in the tap's root directory or can nest them under an e.g. `formula_lists/` directory. I'm leaning toward the second option because I think it is cleaner and adding more lists in the future won't increase clutter.

Given the above: I'd be tempted to put them in either files or directories relating to the command(s) that will run them.

> For example, `tap.formula_lists` may return:

👍🏻 

> However, none of the questions listed above will affect this PR, so I'm going to push forward with this (modifying to keep the audit directory in `audit_exceptions/`)

👍🏻

**Comment by Rylan12** [comment]
> > * We will keep using `audit_exceptions/`
> 
> 👍🏻
> 
> > * We can either add these to the existing `audit_exceptions/` directory or create a new `rubocop_exceptions/` or `style_exceptions/` directory. I'm leaning toward `style_exceptions/` because it matches with the command (`brew style` not `brew rubocop`) so it's clearer to anyone who doesn't know about rubocop or the inner workings of `brew style`.
> 
> 👍🏻 to either `audit_exceptions` or `style_exceptions`.
> 
> > * We can either add these as individual files in the tap's root directory or can nest them under an e.g. `formula_lists/` directory. I'm leaning toward the second option because I think it is cleaner and adding more lists in the future won't increase clutter.
> 
> Given the above: I'd be tempted to put them in either files or directories relating to the command(s) that will run them.
> 
> > For example, `tap.formula_lists` may return:
> 
> 👍🏻
> 
> > However, none of the questions listed above will affect this PR, so I'm going to push forward with this (modifying to keep the audit directory in `audit_exceptions/`)
> 
> 👍🏻

Okay. May as well discuss this in #9185. For now, is this PR good to go? If so, I'll start the merge. I don't think we'll need to wait for linuxbrew-core merges as this change isn't breaking CI.

**Review by MikeMcQuaid — APPROVED** [review]

**Pull request merged**
